src/projection.jl

```diff
@@ -0,0 +1,77 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    projector([T::Type], x)
+
+Returns a `project(dx)` closure which maps `dx` onto type `T`, such that it is the
+same size as `x`. If `T` is not provided, it is assumed to be the type of `x`.
+
+It's necessary to have `x` to ensure that it's possible to project e.g. `AbstractZero`s
+onto `Array`s -- this wouldn't be possible with type information alone because the neither
+`AbstractZero`s nor `T` know what size of `Array` to produce.
+"""
+function projector end
+
+projector(x) = projector(typeof(x), x)
+
+# fallback (structs)
+function projector(::Type{T}, x::T) where T
+    project(dx::T) = dx
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Numbers
+function projector(::Type{T}, x::T) where {T<:Real}
+    project(dx::Real) = T(dx)
+    project(dx::Number) = T(real(dx)) # to avoid InexactError
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+function projector(::Type{T}, x::T) where {T<:Number}
+    project(dx::Number) = T(dx)
+    project(dx::AbstractZero) = zero(x)
+    project(dx::AbstractThunk) = project(unthunk(dx))
+    return project
+end
+
+# Arrays
+function projector(::Type{Array{T, N}}, x::Array{T, N}) where {T, N}
+    sizex = size(x)
+    projT = projector(zero(T))
+    project(dx::Array{T, N}) = dx # identity
+    project(dx::AbstractArray) = project(collect(dx)) # from Diagonal
```

Comment:
Here I also wonder if this is the right behaviour. Maybe the ability to reproduce a similar dense array is desirable sometimes, but making the default projector materialise when it doesn't have to seems odd --- shouldn't we preserve Diagonal or Fill backwards as many steps as possible, by default? 

But again maybe this is trying to serve multiple purposes which perhaps can be clarified.

Reply:
Maybe there ought to be abstract types involved, something like:
```
projector(x::Real) = projector(Real, x)
projector(x::Bool) = projector(Nothing, x)

projector(x::AbstractArray{<:Real}) = projector(AbstractArray{Real}, x)
projector(x::AbstractArray) = projector(AbstractArray, x)
```
where `projector(AbstractArray, x)(dx)` may reshape but won't do more.

Reply:
This is the method which specifically wants the output to be a dense array, i.e. where `x` is a `Matrix` in `projector(x)` call. When `x` is a `Diagonal`, a different `projector` method would be hit.

I couldn't quite see how to generalise the method for an arbitrary `AbstractArray` (see how `Diagonal` and `Symmetric`) cases are different. My plan was to just add the dispatch for any type that we need to make ChainRules rules work.